### PR TITLE
do not process event stream if it is empty to avoid producing a stack…

### LINKED ISSUE #116: emit transaction failed: error_class=NoMethodError error="undefined method `[]' for nil:NilClass"
Hey Folks:

First of all, thanks for building a fantastic plugin! We pair your plugin with our own to get the data from k8s into our platform  We also leverage the [Concat plugin](https://github.com/fluent-plugins-nursery/fluent-plugin-concat) to handle multiline line logging.  We discovered a bug with our configuration of the concat plugin, however after addressing, we started seeing some errors from the k8s metadata plugun:

```
fluentd-2tl8n fluentd 2018-04-01 18:44:35 +0000 [warn]: #0 emit transaction failed: error_class=NoMethodError error="undefined method `[]' for nil:NilClass" tag="containers.mnt.log.containers.calico-node-hxmb6_kube-system_calico-node-703da71b5a5aa7a72c4eef2653357809f11cb03bfb6fab74daa7e9db92b4acc3.log"
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluent-plugin-kubernetes_metadata_filter-1.0.1/lib/fluent/plugin/filter_kubernetes_metadata.rb:297:in `filter_stream_from_files'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/event_router.rb:177:in `block in filter_stream'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/event_router.rb:177:in `each'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/event_router.rb:177:in `reduce'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/event_router.rb:177:in `filter_stream'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/event_router.rb:158:in `emit_events'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/event_router.rb:96:in `emit_stream'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/plugin/out_relabel.rb:29:in `process'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/plugin/output.rb:721:in `emit_sync'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/event_router.rb:159:in `emit_events'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/event_router.rb:96:in `emit_stream'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/plugin/in_tail.rb:348:in `receive_lines'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/plugin/in_tail.rb:465:in `wrap_receive_lines'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/plugin/in_tail.rb:666:in `block in on_notify'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/plugin/in_tail.rb:707:in `with_io'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/plugin/in_tail.rb:644:in `on_notify'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/plugin/in_tail.rb:496:in `on_notify'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/plugin_helper/timer.rb:77:in `on_timer'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/cool.io-1.5.0/lib/cool.io/loop.rb:88:in `run_once'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/cool.io-1.5.0/lib/cool.io/loop.rb:88:in `run'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/plugin_helper/event_loop.rb:84:in `block in start'
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 /var/lib/gems/2.3.0/gems/fluentd-0.14.17/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
fluentd-2tl8n fluentd 2018-04-01 18:44:35 +0000 [warn]: #0 emit transaction failed: error_class=NoMethodError error="undefined method `[]' for nil:NilClass" tag="containers.mnt.log.containers.calico-node-hxmb6_kube-system_calico-node-703da71b5a5aa7a72c4eef2653357809f11cb03bfb6fab74daa7e9db92b4acc3.log"
fluentd-2tl8n fluentd   2018-04-01 18:44:35 +0000 [warn]: #0 suppressed same stacktrace
```

I dug in further and discovered that the stack was coming from the `filter_stream_from_files` method:

https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter/blob/master/lib/fluent/plugin/filter_kubernetes_metadata.rb#L297

The stack occurs when the es is empty.  I tested a patch locally by checking if the es is empty before continuing.  Happy to submit a PR on this but wanted to check on the logic of the issue and patch before doing so.

### PR DESCRIPTION
**Pull request by frankreno**
fixes #116 by checking if the event stream is empty before continuing to process...

### CODE REVIEW

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 1]
```diff
@@ -284,67 +284,71 @@ def filter_stream(tag, es)
     end
 
     def filter_stream_from_files(tag, es)
-      new_es = MultiEventStream.new
```
Comment (richm) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: how about a short circuit like this:
```
    if es.empty?
      return es
    end
```
Same with filter_stream_from_journal

Also, please add a test for this.

Thanks!

**Comment by frankreno** [comment]
@richm certainly.  Apologies for the noob question.  I am struggling to see how to emit an empty EventStream in the test class that would trigger the short circuit. Can you give me any pointers?

File: lib/fluent/plugin/filter_kubernetes_metadata.rb (hunk 1 above)
Comment (frankreno) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: @richm certainly. Apologies for the noob question. I am struggling to see how to emit an empty EventStream in the test class that would trigger the short circuit. Can you give me any pointers?
--



@richm certainly. Apologies for the noob question. I am struggling to see how to emit an empty EventStream in the test class that would trigger the short circuit. Can you give me any pointers?
--

Comment (richm) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: not sure - MultiEventStream.new creates an empty stream?

Comment (frankreno) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: Yes correct, I am just trying to figure out how to call emit() and have it return an empty stream:

```
def emit(record, time = Engine.now)
  emit_with_tag(@tag, record, time)
end

def emit_with_tag(tag, record, time = Engine.now)
  @events[tag] ||= MultiEventStream.new
  @events[tag].add(time, record)
end
```
`emit_with_tag` always creates a new stream and adds the record.  So I cannot see how to make it empty.

Comment (richm) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: hmm - I have no idea - which makes me wonder how it is possible to run into this situation in the first place . . .

**Comment by frankreno** [comment]
That code I shared was from the filter test class, not the core code.

File: lib/fluent/plugin/filter_kubernetes_metadata.rb (hunk 1 above)
Comment (richm) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: unless you can figure out how to test this, I guess we just won't have a test for this case - I'm not sure how to test with an empty event stream

so just put the short circuit return in the filter_stream_from_files method.  Looks like filter_stream_from_journal is ok

Comment (frankreno) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: I beat my head against the wall trying to find a way to test this but I can't.  At least this adds a safety check to ensure we do not do any invalid options.  Updated with the requested changes.  You were right, journal did not need a change at all.

**Comment by richm** [comment]
looks good - please squash the commits into a single commit and resubmit and I'll merge

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 2]
```diff
@@ -284,71 +284,70 @@ def filter_stream(tag, es)
...
-        if match_data
-          container_id = match_data['docker_id']
-          metadata = {
-            'docker' => {
-              'container_id' => container_id
-            },
-            'kubernetes' => get_metadata_for_record(match_data, container_id, create_time_from_record(es.first[1]), batch_miss_cache)
-          }
-        end
+      if es.empty?
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: style: `return es if es.empty?`

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 3]
```diff
@@ -284,71 +284,70 @@ def filter_stream(tag, es)
...
-        new_es
-      end
+        new_es.add(time, record)
+      }
+      dump_stats
+      new_es
     end
 
     def filter_stream_from_journal(tag, es)
-      unless es.empty?
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: We seem to have lost the empty check here

Comment (richm) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: We don't need it here - the code already correctly handles empty es.

**Comment by jcantrill** [comment]
@frankreno please squash into a single commit

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 4]
```diff
@@ -313,6 +313,9 @@ def filter_stream_from_files(tag, es)
     end
 
     def filter_stream_from_journal(tag, es)
+      if es.empty?
```
Comment (richm) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: ruby style - single line: `return es if es.empty?`

Comment (frankreno) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: @richm will do.  Did I correctly squash it into a single commit? Is the right way to do that as I did above or should i just submit a new clean PR with only that change?

Comment (richm) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: typically the way it works is that you use `git rebase -i origin/master` - keep the first commit, and fixup or squash the rest - then `git push -f frankreno master` - that will force push the rebased commit, which will also update this PR

**Comment by frankreno** [comment]
i have gotten to a bad state, let me close this and resubmit it..

**Pull request closed without merging**